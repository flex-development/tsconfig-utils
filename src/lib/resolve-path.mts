/**
 * @file resolvePath
 * @module tsconfig-utils/lib/resolvePath
 */

import withTrailingSlash from '#internal/with-trailing-slash'
import isResolvedTsconfig from '#lib/is-resolved-tsconfig'
import * as mlly from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import type { Tsconfig } from '@flex-development/tsconfig-types'
import type { ResolvePathOptions } from '@flex-development/tsconfig-utils'

/**
 * Resolve an aliased `specifier`.
 *
 * @see {@linkcode ResolvePathOptions}
 *
 * @this {void}
 *
 * @param {string} specifier
 *  The module specifier to resolve
 * @param {ResolvePathOptions | null | undefined} [options]
 *  Alias resolution options
 * @return {string | null}
 *  Resolved specifier or `null` if path alias match is not found
 */
function resolvePath(
  this: void,
  specifier: string,
  options?: ResolvePathOptions | null | undefined
): string | null {
  /**
   * URL of directory to resolve non-absolute modules from.
   *
   * @var {mlly.ModuleId | null | undefined} baseUrl
   */
  let baseUrl: mlly.ModuleId | null | undefined

  /**
   * URL of parent module.
   *
   * @var {mlly.ModuleId | null | undefined} parent
   */
  let parent: mlly.ModuleId | null | undefined

  /**
   * Tsconfig object.
   *
   * @var {Tsconfig | null | undefined} tsconfig
   */
  let tsconfig: Tsconfig | null | undefined

  if (isResolvedTsconfig(options?.tsconfig)) {
    tsconfig = options.tsconfig.tsconfig
    baseUrl = new URL(pathe.dot, options.tsconfig.url)
    parent = options.tsconfig.url
  } else if (options) {
    tsconfig = options.tsconfig
    baseUrl = options.cwd
  }

  if (typeof tsconfig?.compilerOptions?.baseUrl === 'string') {
    baseUrl = tsconfig.compilerOptions.baseUrl
  }

  if (baseUrl !== null && baseUrl !== undefined && baseUrl !== parent) {
    baseUrl = withTrailingSlash(mlly.toUrl(baseUrl))
  }

  return mlly.resolveAlias(specifier, {
    ...options,
    aliases: tsconfig?.compilerOptions?.paths,
    cwd: baseUrl,
    parent: parent ?? baseUrl
  })
}

export default resolvePath
