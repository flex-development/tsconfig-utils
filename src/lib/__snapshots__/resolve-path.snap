// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/resolvePath > should return resolved specifier ("#internal/with-trailing-slash") 1`] = `"./src/internal/with-trailing-slash"`;

exports[`unit:lib/resolvePath > should return resolved specifier ("#lib/is-resolved-tsconfig") 1`] = `"./src/lib/is-resolved-tsconfig"`;

exports[`unit:lib/resolvePath > should return resolved specifier ("https://esm.sh/@flex-development/unist-util-builder") 1`] = `"./typings/@flex-development/unist-util-builder/index.d.mts"`;

exports[`unit:lib/resolvePath > should return resolved specifier ("https://esm.sh/lodash@4.17.21") 1`] = `"../node_modules/@types/lodash/index.d.ts"`;
